Translate emails to recipient's preferred locale

momo3404 opened the pull request:
# Changes Proposed in this PR:

## Added the following methods to `LocaleService`:
 - `def with_preferred_locale(recipient, &)`
   - Localizes the given block to the recipient's preferred locale.
   - `recipient` may be an instance of User or Org (is able to respond to `language&.abbreviation`)
   - Fallback to to the current locale (`I18n.locale`), if necessary.
 - `def with_each_available_locale`
   - Iterates over each available locale, switching to that locale, and yields the locale to the given block for translation and processing.
   
## Updated locale handling for outgoing emails:
 - The following outgoing emails are now localized to a recipient's preferred locale (via `LocaleService.with_preferred_locale(recipient)`):
   - `UserMailer.question_answered`
   - `UserMailer.sharing_notification`
   - `UserMailer.permissions_change_notification`
   - `UserMailer.plan_access_removed`
   - `UserMailer.feedback_notification`
   - `UserMailer.feedback_complete`
   - `UserMailer.plan_visibility`
   - `UserMailer.new_comment`
   - `UserMailer.admin_privileges`
   - `UserMailer.api_credentials`
   - NOTE: The Devise reset password email has not yet been added here. This should eventually be implemented, but will require some Devise overriding to use the recipient's preferrred locale when localising the email subject.
 
 - The following outgoing emails are now rendered in both English and French via (`LocaleService.with_each_available_locale`):
   - `UserMailer.welcome_notification`
   - Devise's Confirmation Instructions email
   - Devise's Invitation Instructions email (triggered when a non-existing email is added as a plan collaborator
   
   - These three emails are English and French because the recipient has not yet had the opportunity to select their preferred locale within the app. Note that the email subject is also rendered in both English and French for all 3 of these emails (can be verified via the following links):
     - https://github.com/portagenetwork/roadmap/pull/1100/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bL423
     - https://github.com/portagenetwork/roadmap/pull/1100/files#diff-55254045ba774e4c360d7bbfde054665072d2ab49abc3f21add15a37bba65a1cR23
     - https://github.com/portagenetwork/roadmap/pull/1100/files#diff-d3f6ed8387f43b48699fba03e304c682fa7f665c0eb9edfe5d9562b80cc5eca7R237
     - https://github.com/portagenetwork/roadmap/pull/1100/files#diff-b7c9801079388d7d92e11a05faae74843651a70316579aa432c8240e20d767e3R250

## Enabled translations for several strings in several mailer views.

## Added tests for the new mailer localization behaviour, as well as for the new `LocaleService` methods.

momo3404 committed:
Enable correct translation for emails

momo3404 committed:
enable translation for password reset email

momo3404 committed:
Enable translation for admin privileges list

aaronskiba commented on app/mailers/user_mailer.rb:
```diff
@@ -41,14 +41,14 @@ def question_answered(data, user, answer, _options_string)
     @helpdesk_email = helpdesk_email(org: @user.org)
 
-    I18n.with_locale I18n.locale do
+    I18n.with_locale(email_locale(@user)) do
```
Did we want to also apply this to line 23? If so, line 17 (`@email_subject  = format(_('Query or feedback related to %{tool_name}'), tool_name: tool_name)`) should me moved inside the `I18n.with_locale` block.

momo3404 committed:
Enable translation for welcome email and refactor test

momo3404 committed:
Enable translation for invitation emails

momo3404 committed:
Enable translation for confirmation email

momo3404 committed:
Made rubocop happy

aaronskiba commented on config/locales/translation.en-CA.yml:
```diff
@@ -234,7 +234,7 @@ en-CA:
     mailer:
       confirmation_instructions:
-        subject: Confirm your DMP Assistant account
+        subject: Confirm your DMP Assistant account / Validez votre compte Assistant PGD
```
This is good, we'll just have to apply the changes on translation.io to make sure they persist.

aaronskiba commented on app/mailers/user_mailer.rb:
```diff
@@ -41,14 +37,14 @@ def question_answered(data, user, answer, _options_string)
     @helpdesk_email = helpdesk_email(org: @user.org)
 
-    I18n.with_locale I18n.locale do
+    I18n.with_locale(email_locale(@user)) do
```
The recipient here is `data['email']`, but we are translating the email for `@user`. Maybe we should verify whether `@user.email == data['email']`.

aaronskiba replied:
I am seeing the following in `app/views/user_mailer/question_answered.html.erb`:
```ruby
<p>
  <%= raw _('Hello %{recipient_name},') %{ recipient_name: @recipient_name } %>
</p>
<p>
  <%= raw @username + _(" is creating a Data Management Plan and has answered ")
```
Which is assigned as follows in `UserMailer.question_answered`:
```ruby
  def question_answered(data, user, answer, _options_string)
    @user           = user
    @username       = @user.name
    @answer         = answer
    @question_title = @answer.question.text.to_s
    @plan_title     = @answer.plan.title.to_s
    @template_title = @answer.plan.template.title.to_s
    @data           = data
    @recipient_name = @data['name'].to_s
    @message        = @data['message'].to_s
```
So I'm thinking we should make an update like the following within `UserMailer.question_answered`:
```ruby
locale = User.find_by(email: data['email'])
I18n.with_locale(email_locale(locale)) do
```
The safe navigation operators within `MailerHelper.email_locale` will protect us if `User.find_by(email: data['email'])` returns nil.

momo3404 replied:
Nice catch, I added that change.

aaronskiba commented on app/mailers/user_mailer.rb:
```diff
@@ -211,7 +207,9 @@ def api_credentials(api_client)
     @helpdesk_email = helpdesk_email(org: @api_client.org)
 
-    I18n.with_locale I18n.locale do
+    recipient = User.find_by(email: @api_client.contact_email) || @api_client.org
```
I'm worried about this being prone to a `NoMethodError`:
Lets say `User.find_by(email: @api_client.contact_email)` is nil and `@api_client.org` is assigned to `recipient`. `email_locale(recipient)` is then called and performs the following:
`recipient.language&.abbreviation || I18n.locale`.

If `@api_client.org` is in fact an Org object, then there is no problem. However, I don't think we can be assured of that. The following association exists in `class ApiClient`:
```ruby
belongs_to :org, optional: true
```
Thus, `recipient` (i.e. `@api_client.org`) may also be nil and `recipient.language&.abbreviation` would raise a `NoMethodError`.

momo3404 replied:
Good point, I think changing the `email_locale` function to check if `recipient` is nil should help avoid this error

aaronskiba commented on app/models/user.rb:
```diff
@@ -421,7 +421,9 @@ def get_preferences(key)
-    super(options.merge(subject: format(_('A Data Management Plan in %{application_name} has been shared with you'),
+    super(options.merge(subject: format(_('A Data Management Plan in %{application_name} has been shared with you ' \
+                                          '/ Un plan de gestion des données dans %{application_name} a été partagé ' \
+                                          'avec vous'),
```
Both instances of `ApplicationService.application_name` will either be translated to English or French here. It would probably be best to just fully hardcode it. That is, remove `format()` as well as `_()` and just provide the hardcoded string.

aaronskiba replied:
I also wonder if rather than overriding the default Devise email subject like this, we could instead set the line in `config/locales/translation.XX.yml`. (Maybe that's a refactor we could potentially perform when we apply the changes upstream?)

momo3404 replied:
Would changing the subject in the yml files not override it for both invitation emails, not just the one for users not in the db?

aaronskiba commented on app/views/user_mailer/admin_privileges.html.erb:
```diff
@@ -6,7 +6,7 @@
   </p>
   <p>
-    <%= sanitize(@ul_list) %>
+    <%= sanitize(privileges_list(@user)) %>
```
I think this is a good change and we should be able to remove `@ul_list   = privileges_list(@user)` from `UserMailer.admin_privileges` as a result. I included that change in this commit from the prior PR: https://github.com/portagenetwork/roadmap/pull/1094/commits/282b2ca7da24753002cb293195f39aac6218e743

aaronskiba replied:
No worries, but maybe it would’ve been better to cherry-pick that aforementioned commit into this PR. That way we’d retain the original author and keep a bit more traceability in the history. Using git cherry-pick or adding a Co-authored-by tag could be good habits to consider when we’re reusing each other’s code. Not a big thing in this PR, but something to consider going forward.

momo3404 replied:
I actually didn't know about that git feature, I'll be sure to use it in the future.

aaronskiba commented on spec/mailers/user_mailer_spec.rb:
```diff
@@ -5,42 +5,14 @@
 require 'rails_helper'
 
 RSpec.describe UserMailer, type: :mailer do
```
The test update works, but I guess we've never had much test coverage for the files being changed in this PR.

momo3404 replied:
I can add some test coverage for more emails here.

momo3404 committed:
Refactored helper function and invitation subject

momo3404 committed:
Added more test coverage

momo3404 committed:
Added more test coverage

Fixed question_answered and admissions emails

Correct variable name

Fixed question_answered and admissions emails

momo3404 committed:
Fixed question_answered and admissions emails

aaronskiba commented on app/mailers/user_mailer.rb:
```diff
@@ -41,14 +37,16 @@ def question_answered(data, user, answer, _options_string)
     @helpdesk_email = helpdesk_email(org: @user.org)
 
-    I18n.with_locale I18n.locale do
+    locale = User.find_by(email: data['email'])
```
Aw, I'm sorry, I had suggested this code, but am regretting `locale` as the variable name. Something like `recipient` would be better.

momo3404 committed:
Correct variable name

momo3404 committed:
Merge branch 'momo/email-translation-fix' of https://github.com/portagenetwork/roadmap into momo/email-translation-fix

aaronskiba commented on app/mailers/user_mailer.rb:
```diff
@@ -20,12 +19,9 @@ def welcome_notification(user)
     @contact_us     = Rails.application.config.x.organisation.contact_us_url || contact_us_url
     @helpdesk_email = helpdesk_email(org: @user.org)
 
-    I18n.with_locale I18n.locale do
```
This is just a comment regarding the previous code we are changing.
```ruby
I18n.with_locale I18n.locale do
  mail(to: @user.email,
           subject: format(_('Welcome to %{tool_name}'), tool_name: tool_name))
end
```
The above block was redundant. `_()` translates using the current locale in use (i.e. `I18n.locale`).
So prior to our changes, the above code could've previously been refactored to the following:
```ruby
mail(to: @user.email,
           subject: format(_('Welcome to %{tool_name}'), tool_name: tool_name))
```

aaronskiba replied:
```ruby
def email_locale(recipient)
  recipient&.language&.abbreviation || I18n.locale
end
```
At the same time, if the above helper function returns ` I18n.locale`, then the same redundancy is occuring. I might try to create a quick refactor for it and we can decide how to proceed.

aaronskiba commented:
It might be better to title this PR something like "Translate emails to recipient's preferred locale."

aaronskiba commented on spec/mailers/user_mailer_spec.rb:
```diff
@@ -36,12 +53,41 @@
     end
   end
+
+  describe 'Invitation Email to User Not in DB' do
```
It's good you are testing the overriding of the Devise email subject, but maybe this test belongs somewhere else?

momo3404 replied:
Do you think `spec/integration/email_confirmation_spec.rb` is a better location?

aaronskiba replied:
It might be best to create a new file: `spec/mailers/user_invitation_mailer_spec.rb`

aaronskiba commented on spec/mailers/user_mailer_spec.rb:
```diff
@@ -36,12 +53,41 @@
+
+      mail = ActionMailer::Base.deliveries.last
+      expect(mail.to).to include('newuser@example.com')
+      expect(mail.subject).to include('A Data Management Plan in DMP Assistant has been shared with you')
```
Maybe a single `expect(mail.subject).to eq` test that verifies the exact hardcoded string would be best here?

aaronskiba committed:
Refactor I18n.with_locale(email_locale(x))

Extracts logic for determining a recipient's preferred locale into `LocaleService.with_preferred_locale`. This change abstracts `I18n.with_locale(recipient.language.abbreviation)` calls.

aaronskiba committed:
Refactor locale iteration into `with_each_available_locale`

This change introduces `LocaleService.with_each_available_locale`, which encapsulates the common pattern of iterating over `I18n.available_locales` and yielding each locale within `I18n.with_locale`.

By extracting this logic, we reduce repetition and improve readability by eliminating the need for nested locale iteration blocks.

aaronskiba committed:
Refactor conditional <hr> in mailer views

aaronskiba committed:
Add Tests for new LocaleService methods

Adds RSpec tests for the following methods:
- `LocaleService.with_preferred_locale`
- `LocaleServcie.with_each_available_locale`

aaronskiba committed:
Move invitation email tests to separate _spec file

aaronskiba committed:
Update user_invitation_mailer_spec.rb

These changes are intended to do the following:
- Simplify the logic for mocking the delivery of the invitation email
- Add the `invite_new_user` method to more closely emulate the `User.invite!` behaviour within `RolesController#create`
- Add a test to verify that the email body is bilingual

aaronskiba committed:
Fix for sharing_notification string spacing to enable translation

The translation for this string existed on translation.io, however it was not translating. This change simply removes the extra/unnecessary whitespace and newline to avoid translation issues.

aaronskiba committed:
Refactor and add coverage for UserMailer tests

- Consolidated duplicate `UserMailer.sharing_notification` code via `shared_examples`
- Added coverage to also test/verify subject lines and body content of delivered emails
- Set I18n.available_locales temporarily to simulate application environment
- Ensured I18n.locale remains unchanged after mail delivery

momo3404 committed:
Merge pull request #1111 from portagenetwork/aaron/email-translation-fix/refactor

Refactor Email Localization Logic and Add Test Coverage

aaronskiba committed:
Remove `with_preferred_locale` from reset password view

Removing `<% LocaleService.with_preferred_locale(user) do %>` addresses an inconsistency with the delivered reset password instructions email. Specifically, the removed code was setting the email body to the user's preferred language, whereas the reset password instructions email subject was translating according to the app's current language.

Localising the email subject to the user's preferred locale would entail some extra Devise overriding. So for now, to keep things simple and consistent, we are also using the app's current locale to set the language of the email body.

aaronskiba committed:
Update config/locales/*.yml via translation:sync

Ran bundle exec rake translation:sync with `config.disable_yaml` temporarily set to false.

Due to an incompatibility between Ruby 3.1 and its default Psych gem version (https://bugs.ruby-lang.org/issues/17866), translation:sync could only be executed successfully by also temporarily adding `gem 'psych', '~> 3.3.4'` to Gemfile.

momo3404 committed:
Add changelog entry
